File: 0407-trapping-rain-water-ii/0407-trapping-rain-water-ii.cpp
class Solution {
public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        int R= heightMap.size();
        int C = heightMap[0].size();
        vector<vector<int>>vis(R,vector<int>(C,0));
        int vol=0;
        //by deault pr in c++ is max heap so we have to declare greater and container for pq is vector
        priority_queue<pair<int,pair<int,int>>, vector<pair<int,pair<int,int>>>,greater<pair<int,pair<int,int>>>>pq;
        for(int i=0;i<R;i++){
            for(int j=0;j<C;j++){
                if(i==0 ||i==R-1 ||j==0||j==C-1){
                    pq.push({heightMap[i][j],{i,j}});
                    vis[i][j]=true;
                }
            }
        }
        int minBdHt=0;
        int dR[4]={0,0,1,-1};
        int dC[4]={1,-1,0,0};
        while(!pq.empty()){
            int curHt = pq.top().first;
            int curR = pq.top().second.first;
            int curC = pq.top().second.second;
            minBdHt = max(minBdHt, curHt);
            pq.pop();
            for(int i=0;i<4;i++){
                int tR = curR+dR[i];
                int tC =  curC +dC[i];
                if(tR >=0 && tR<R && tC >=0 && tC <C && vis[tR][tC]==false){
                    vis[tR][tC]=true;
                    pq.push({heightMap[tR][tC],{tR,tC}});
                    if(heightMap[tR][tC]<minBdHt){
                        vol += minBdHt-heightMap[tR][tC];
                    }
                }
            }
        }
        return vol;
    }
};

File: 0407-trapping-rain-water-ii/0407-trapping-rain-water-ii.java
class Solution {
    public int trapRainWater(int[][] heightMap) {
        if (heightMap == null || heightMap.length == 0 || heightMap[0].length == 0) {
            return 0;
        }

        int n = heightMap.length;
        int m = heightMap[0].length;

        // Min-Heap to store cells with their heights
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        boolean[][] vis = new boolean[n][m];

        // Add boundary cells to the priority queue
        for (int i = 0; i < n; i++) {
            pq.offer(new int[] {heightMap[i][0], i, 0});
            pq.offer(new int[] {heightMap[i][m - 1], i, m - 1});
            vis[i][0] = true;
            vis[i][m - 1] = true;
        }
        for (int i = 0; i < m; i++) {
            pq.offer(new int[] {heightMap[0][i], 0, i});
            pq.offer(new int[] {heightMap[n - 1][i], n - 1, i});
            vis[0][i] = true;
            vis[n - 1][i] = true;
        }

        int[] dR = {0, 0, 1, -1};
        int[] dC = {1, -1, 0, 0};

        int vol = 0;
        int minHt = 0;

        // Process cells in the priority queue
        while (!pq.isEmpty()) {
            int[] cell = pq.poll();
            int height = cell[0];
            int r = cell[1];
            int c = cell[2];

            // Update the minimum height encountered so far
            minHt = Math.max(minHt, height);

            // Check all 4 neighbors
            for (int i = 0; i < 4; i++) {
                int tR = r + dR[i];
                int tC = c + dC[i];

                if (tR >= 0 && tR < n && tC >= 0 && tC < m && !vis[tR][tC]) {
                    vis[tR][tC] = true;
                    int neighborHeight = heightMap[tR][tC];
                    // Add water trapped at this cell
                    if (neighborHeight < minHt) {
                        vol += minHt - neighborHeight;
                    }
                    // Add the neighbor cell to the queue
                    pq.offer(new int[] {Math.max(neighborHeight, minHt), tR, tC});
                }
            }
        }

        return vol;
    }
}
